File: src/core/expeditions.ts
/**
 * Expediciones: la criatura sale a buscar cosas y vuelve más tarde.
 *
 * Son las dos mitades que le faltaban al juego:
 *
 * 1. El lado de la OFERTA de la economía. Con inventario pero sin forma de
 *    conseguir comida, quedarse sin nada sería un callejón sin salida.
 * 2. Un motivo concreto para volver mañana. Algo que corre mientras no estás y
 *    te espera cuando volvés.
 *
 * ---
 *
 * LA REGLA QUE EVITA EL BLOQUEO: el patio siempre está disponible.
 *
 * Es la trampa clásica de una economía cerrada: si la comida se acaba y para
 * conseguir más hace falta comida, el jugador queda trabado sin nada que hacer.
 * El patio no pide etapa, no cuesta energía y siempre trae algo. Es aburrido a
 * propósito — es un piso, no una estrategia. Los destinos que valen la pena
 * piden criatura crecida y energía.
 */

import type { Stage } from "./evolution.ts";
import type { Inventario } from "./inventory.ts";
import { deriveSeed, mulberry32 } from "./rng.ts";
// `Expedicion` se declara junto al estado de la criatura, no acá: si viviera en
// este módulo, simulation.ts tendría que importarlo y quedaría un ciclo.
import type { CreatureState, Expedicion } from "./simulation.ts";

export type { Expedicion };

const MINUTO = 60_000;
const HORA = 60 * MINUTO;

export interface Destino {
  id: string;
  nombre: string;
  /**
   * Cómo se dice "volvió de ___".
   *
   * Va escrito y no armado concatenando: "de" + "el patio" da "de el patio".
   * La contracción del castellano no sale de pegar strings.
   */
  desde: string;
  descripcion: string;
  duracionMs: number;
  costoEnergia: number;
  etapaMinima: Stage;
  /** Cuántos alimentos trae, mínimo y máximo. */
  items: readonly [number, number];
  /** Peso relativo de cada alimento en el sorteo. */
  pesos: Readonly<Record<string, number>>;
  /** Probabilidad de volver con una semilla. */
  chanceSemilla: number;
}

const ORDEN_ETAPA: Record<Stage, number> = { bebe: 0, juvenil: 1, adulto: 2 };

export const DESTINOS: readonly Destino[] = [
  {
    id: "patio",
    nombre: "El patio",
    desde: "del patio",
    descripcion: "Da una vuelta por acá nomás. Siempre trae algo.",
    duracionMs: 15 * MINUTO,
    // Sin costo y sin etapa mínima: es la salida que evita quedar trabado.
    costoEnergia: 0,
    etapaMinima: "bebe",
    items: [1, 2],
    pesos: { baya: 3, raiz: 2 },
    chanceSemilla: 0,
  },
  {
    id: "bosque",
    nombre: "El bosque",
    desde: "del bosque",
    descripcion: "Un rato largo entre los árboles. A veces encuentra semillas.",
    duracionMs: 90 * MINUTO,
    costoEnergia: 15,
    etapaMinima: "juvenil",
    items: [2, 3],
    pesos: { larva: 3, raiz: 3, baya: 2 },
    chanceSemilla: 0.18,
  },
  {
    id: "ruinas",
    nombre: "Las ruinas",
    desde: "de las ruinas",
    descripcion: "Lejos y pesado. Vuelve con cosas que no se ven en otro lado.",
    duracionMs: 4 * HORA,
    costoEnergia: 30,
    etapaMinima: "adulto",
    items: [3, 5],
    pesos: { cristal: 2, larva: 3, raiz: 2, baya: 1 },
    chanceSemilla: 0.55,
  },
];

export function destinoPorId(id: string): Destino | undefined {
  return DESTINOS.find((d) => d.id === id);
}

export interface Botin {
  alimentos: Inventario;
  /** Genoma encontrado, para incubar. `null` si no trajo ninguno. */
  semilla: bigint | null;
}

export type PuedeSalir = { puede: true } | { puede: false; motivo: string };

export function estaFuera(criatura: CreatureState): boolean {
  return criatura.expedicion !== null;
}

export function puedeSalir(criatura: CreatureState, destino: Destino, _nowMs: number): PuedeSalir {
  if (estaFuera(criatura)) return { puede: false, motivo: "Ya está afuera." };
  if (criatura.letargico) return { puede: false, motivo: "Está en letargo." };

  if (ORDEN_ETAPA[criatura.etapa] < ORDEN_ETAPA[destino.etapaMinima]) {
    return { puede: false, motivo: "Todavía está muy chica para ir tan lejos." };
  }
  if (criatura.stats.energia < destino.costoEnergia) {
    return { puede: false, motivo: "No le da la energía. Primero que coma algo." };
  }
  return { puede: true };
}

/** Manda a la criatura. No muta la que recibe. */
export function enviar(criatura: CreatureState, destino: Destino, nowMs: number): CreatureState {
  return {
    ...criatura,
    stats: {
      ...criatura.stats,
      energia: Math.max(0, criatura.stats.energia - destino.costoEnergia),
    },
    expedicion: {
      destinoId: destino.id,
      salidaMs: nowMs,
      regresoMs: nowMs + destino.duracionMs,
    },
  };
}

export function yaVolvio(criatura: CreatureState, nowMs: number): boolean {
  return criatura.expedicion !== null && nowMs >= criatura.expedicion.regresoMs;
}

/** Cuánto falta para que vuelva, en milisegundos. */
export function faltaParaVolver(criatura: CreatureState, nowMs: number): number {
  if (!criatura.expedicion) return 0;
  return Math.max(0, criatura.expedicion.regresoMs - nowMs);
}

/**
 * Qué trae de vuelta.
 *
 * Determinista a partir del genoma y del momento de salida, así que el botín
 * queda decidido cuando sale, no cuando volvés a mirar. Sin eso, cerrar y
 * reabrir hasta que salga un cristal sería la estrategia óptima.
 */
export function resolverBotin(seed: bigint, destino: Destino, salidaMs: number): Botin {
  const rng = mulberry32(deriveSeed(seed, `expedicion:${destino.id}:${salidaMs}`));

  const opciones = Object.entries(destino.pesos);
  const pesoTotal = opciones.reduce((suma, [, peso]) => suma + peso, 0);

  const [minimo, maximo] = destino.items;
  const cantidad = rng.range(minimo, maximo);

  const alimentos: Inventario = {};
  for (let i = 0; i < cantidad; i++) {
    let tirada = rng.next() * pesoTotal;
    for (const [id, peso] of opciones) {
      tirada -= peso;
      if (tirada <= 0) {
        alimentos[id] = (alimentos[id] ?? 0) + 1;
        break;
      }
    }
  }

  // Semilla de 64 bits armada con dos tiradas de 32.
  const semilla =
    rng.next() < destino.chanceSemilla
      ? (BigInt(rng.int(0x100000000)) << 32n) | BigInt(rng.int(0x100000000))
      : null;

  return { alimentos, semilla };
}

export interface Regreso {
  criatura: CreatureState;
  destino: Destino;
  botin: Botin;
}

/**
 * Recibe a la criatura si ya volvió.
 *
 * Devuelve `null` si sigue afuera o si nunca salió. Volver cuenta como
 * atención: estuvo trabajando para vos, no abandonada.
 */
export function recibir(criatura: CreatureState, nowMs: number): Regreso | null {
  const expedicion = criatura.expedicion;
  if (!expedicion || nowMs < expedicion.regresoMs) return null;

  const destino = destinoPorId(expedicion.destinoId);
  if (!destino) {
    // Un destino que ya no existe (por ejemplo, tras una actualización del
    // juego). Se la trae de vuelta con las manos vacías en vez de dejarla
    // atrapada afuera para siempre.
    return {
      criatura: { ...criatura, expedicion: null, ticksSinCuidado: 0 },
      destino: DESTINOS[0] as Destino,
      botin: { alimentos: {}, semilla: null },
    };
  }

  return {
    criatura: { ...criatura, expedicion: null, ticksSinCuidado: 0 },
    destino,
    botin: resolverBotin(BigInt(criatura.seed), destino, expedicion.salidaMs),
  };
}

/** Resume el botín en una frase. */
export function describirBotin(botin: Botin, nombres: Record<string, string>): string {
  const partes = Object.entries(botin.alimentos).map(
    ([id, cantidad]) => `${cantidad} ${nombres[id]?.toLowerCase() ?? id}`,
  );
  if (partes.length === 0 && !botin.semilla) return "Volvió con las manos vacías.";

  const comida = partes.length > 0 ? `Trajo ${partes.join(", ")}.` : "No trajo comida.";
  return botin.semilla ? `${comida} Y una semilla desconocida.` : comida;
}

File: gdext/src/expeditions.cpp
/**
 * expeditions.cpp — ver expeditions.h.
 */

#include "expeditions.h"
#include "actions.h"
#include "rng.h"

#include <algorithm>

namespace petbits {

static constexpr int64_t MINUTO = 60'000;
static constexpr int64_t HORA = 60 * MINUTO;

const std::vector<Destino>& destinos() {
    static const std::vector<Destino> lista = {
        {
            "patio", "El patio", "del patio",
            "Da una vuelta por acá nomás. Siempre trae algo.",
            15 * MINUTO,
            // Sin costo y sin etapa mínima: es la salida que evita quedar trabado.
            0.0, Stage::Bebe,
            1, 2,
            {{"baya", 3}, {"raiz", 2}},
            0.0,
        },
        {
            "bosque", "El bosque", "del bosque",
            "Un rato largo entre los árboles. A veces encuentra semillas.",
            90 * MINUTO,
            15.0, Stage::Juvenil,
            2, 3,
            {{"larva", 3}, {"raiz", 3}, {"baya", 2}},
            0.18,
        },
        {
            "ruinas", "Las ruinas", "de las ruinas",
            "Lejos y pesado. Vuelve con cosas que no se ven en otro lado.",
            4 * HORA,
            30.0, Stage::Adulto,
            3, 5,
            {{"cristal", 2}, {"larva", 3}, {"raiz", 2}, {"baya", 1}},
            0.55,
        },
    };
    return lista;
}

const Destino* destinoPorId(std::string_view id) {
    for (const Destino& d : destinos()) {
        if (d.id == id) return &d;
    }
    return nullptr;
}

bool estaFuera(const CreatureState& criatura) {
    return criatura.expedicion.has_value();
}

PuedeSalir puedeSalir(const CreatureState& criatura, const Destino& destino) {
    if (estaFuera(criatura)) return {false, "Ya está afuera."};
    if (criatura.letargico) return {false, "Está en letargo."};

    if (static_cast<int>(criatura.etapa) < static_cast<int>(destino.etapaMinima)) {
        return {false, "Todavía está muy chica para ir tan lejos."};
    }
    if (criatura.stats.energia < destino.costoEnergia) {
        return {false, "No le da la energía. Primero que coma algo."};
    }
    return {true, ""};
}

CreatureState enviar(const CreatureState& criatura, const Destino& destino, int64_t nowMs) {
    CreatureState next = criatura;
    next.stats.energia = std::max(0.0, next.stats.energia - destino.costoEnergia);
    next.expedicion = Expedicion{std::string(destino.id), nowMs, nowMs + destino.duracionMs};
    return next;
}

bool yaVolvio(const CreatureState& criatura, int64_t nowMs) {
    return criatura.expedicion.has_value() && nowMs >= criatura.expedicion->regresoMs;
}

int64_t faltaParaVolver(const CreatureState& criatura, int64_t nowMs) {
    if (!criatura.expedicion.has_value()) return 0;
    return std::max<int64_t>(0, criatura.expedicion->regresoMs - nowMs);
}

Botin resolverBotin(Seed seed, const Destino& destino, int64_t salidaMs) {
    // La etiqueta lleva el destino y el momento de salida, así que dos salidas
    // al mismo lugar en momentos distintos traen cosas distintas, y la misma
    // salida siempre trae lo mismo.
    const std::string etiqueta =
        "expedicion:" + std::string(destino.id) + ":" + std::to_string(salidaMs);
    Rng rng = rngFor(seed, etiqueta);

    int pesoTotal = 0;
    for (const PesoAlimento& p : destino.pesos) pesoTotal += p.peso;

    const int64_t cantidad = rng.rango(destino.itemsMin, destino.itemsMax);

    Botin botin;
    for (int64_t i = 0; i < cantidad; ++i) {
        double tirada = rng.next() * pesoTotal;
        for (const PesoAlimento& p : destino.pesos) {
            tirada -= p.peso;
            if (tirada <= 0.0) {
                botin.alimentos.agregar(p.id, 1);
                break;
            }
        }
    }

    // Semilla de 64 bits armada con dos tiradas de 32.
    //
    // El orden de las dos llamadas importa y no es intercambiable: en JS el
    // operando izquierdo de `<<` se evalúa primero, así que la tirada alta sale
    // antes que la baja. Escribirlo en una sola expresión en C++ sería un error
    // —el orden de evaluación de los operandos no está especificado— así que van
    // en dos líneas.
    if (rng.next() < destino.chanceSemilla) {
        const uint64_t alta = static_cast<uint64_t>(rng.intMenorQue(0x100000000LL));
        const uint64_t baja = static_cast<uint64_t>(rng.intMenorQue(0x100000000LL));
        botin.semilla = (alta << 32) | baja;
    }

    return botin;
}

Regreso recibir(const CreatureState& criatura, int64_t nowMs) {
    Regreso r{};
    r.volvio = false;
    r.destino = nullptr;

    if (!criatura.expedicion.has_value() || nowMs < criatura.expedicion->regresoMs) {
        return r;
    }

    const std::string destinoId = criatura.expedicion->destinoId;

    r.volvio = true;
    r.criatura = criatura;
    r.criatura.expedicion = std::nullopt;
    r.criatura.ticksSinCuidado = 0;

    const Destino* destino = destinoPorId(destinoId);
    if (destino == nullptr) {
        // Un destino que ya no existe, por ejemplo tras una actualización del
        // juego. Se la trae de vuelta con las manos vacías en vez de dejarla
        // atrapada afuera para siempre.
        r.destino = &destinos()[0];
        return r;
    }

    r.destino = destino;
    r.botin = resolverBotin(criatura.seed, *destino, criatura.expedicion->salidaMs);
    return r;
}

std::string describirBotin(const Botin& botin) {
    std::string partes;
    for (const auto& [id, cantidad] : botin.alimentos.items()) {
        if (cantidad <= 0) continue;
        if (!partes.empty()) partes += ", ";
        partes += std::to_string(cantidad) + " ";
        // El nombre en minúscula sale de la tabla y no se baja de caso acá: la
        // "í" de "Raíz" ocupa dos bytes en UTF-8 y tolower sobre bytes la rompe.
        const Food* f = buscarAlimento(id);
        partes += f != nullptr ? std::string(f->nombreMinuscula) : id;
    }

    if (partes.empty() && !botin.semilla.has_value()) return "Volvió con las manos vacías.";

    const std::string comida = partes.empty() ? "No trajo comida." : "Trajo " + partes + ".";
    return botin.semilla.has_value() ? comida + " Y una semilla desconocida." : comida;
}

} // namespace petbits
